signature/jws/jws.go

```diff
@@ -25,24 +35,374 @@ func NewEnvelope() signature.Envelope {
+	err = verifyJWT(compact, cert.PublicKey)
+	if err != nil {
+		return nil, nil, err
+	}
+	// parse payload
+	payloadContent, err := base64.RawURLEncoding.DecodeString(e.internalEnvelope.Payload)
+	if err != nil {
+		return nil, nil, err
+	}
+	signerInfo, err := e.SignerInfo()
+	if err != nil {
+		return nil, nil, err
+	}
+	return &signature.Payload{
+		Content:     payloadContent,
+		ContentType: signature.MediaTypePayloadV1}, signerInfo, nil
 }
 
 func (e *envelope) Payload() (*signature.Payload, error) {
-	return nil, fmt.Errorf("not implemented")
+	return &signature.Payload{
+		Content:     []byte(e.internalEnvelope.Payload),
+		ContentType: signature.MediaTypePayloadV1}, nil
 }
 
 func (e *envelope) SignerInfo() (*signature.SignerInfo, error) {
-	return nil, fmt.Errorf("not implemented")
+	signInfo := signature.SignerInfo{}
+	if e.internalEnvelope == nil {
+		return nil, signature.SignatureNotFoundError{}
+	}
+
+	// parse protected headers
+	protected, err := parseProtectedHeaders(e.internalEnvelope.Protected)
+	if err != nil {
+		return nil, err
+	}
+	if err := populateProtectedHeaders(protected, &signInfo); err != nil {
+		return nil, err
+	}
+
+	// parse signature
+	sig, err := base64.RawURLEncoding.DecodeString(e.internalEnvelope.Signature)
+	if err != nil {
+		return nil, err
+	}
+	signInfo.Signature = sig
+
+	// parse headers
+	var certs []*x509.Certificate
+	for _, certBytes := range e.internalEnvelope.Header.CertChain {
+		cert, err := x509.ParseCertificate(certBytes)
+		if err != nil {
+			return nil, err
+		}
+		certs = append(certs, cert)
+	}
+	signInfo.CertificateChain = certs
+	signInfo.UnsignedAttributes.SigningAgent = e.internalEnvelope.Header.SigningAgent
+	signInfo.TimestampSignature = e.internalEnvelope.Header.TimestampSignature
+
+	return &signInfo, nil
+}
+
+// sign the given payload and headers using the given signing method and signature provider
+func sign(payload []byte, headers map[string]interface{}, signer signature.Signer) (string, []*x509.Certificate, error) {
+	jsonPHeaders, err := json.Marshal(headers)
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to encode protected headers: %v", err)
+	}
+	protectedRaw := base64.RawURLEncoding.EncodeToString(jsonPHeaders)
+	payloadRaw := base64.RawURLEncoding.EncodeToString(payload)
+	digest := protectedRaw + "." + payloadRaw
+
+	// check external or internal signer
+
+	sigB, err := signer.Sign([]byte(digest))
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to sign digest. error : %v", err)
+	}
+	finalSig := digest + "." + base64.RawURLEncoding.EncodeToString(sigB)
+
+	certs, err := signer.CertificateChain()
+	if err != nil {
+		return "", nil, err
+	}
+	return finalSig, certs, err
+}
+
+func validateCertificateChain(certChain []*x509.Certificate, expectedAlg signature.Algorithm, f func(string) error) error {
+	if len(certChain) == 0 {
+		return f("certificate-chain not present or is empty")
+	}
+
+	err := validateCertChain(certChain)
```

Comment:
Do we need to implement the cert chain check details again? see https://github.com/notaryproject/notation-core-go/blob/0553376c4e5d86681c177cbb5bfec37024dd69f6/signature/internal/base/envelope.go#L172

Reply:
Thanks! Updated.